serve/mlc_serve/engine/staging_engine_worker.py

```diff
@@ -293,15 +295,21 @@ def _adjust_batch(self):
                 state = self.queue[0]
                 num_tokens = len(state.token_ids)
                 num_new_batched_tokens += num_tokens
+                # this can happen if we processed some request and then had to remove it from inference
+                # due to limit of the kv cache. It can appear that new size to process is bigger
+                # than max_num_batched_tokens. Need to roll back to acceptable size
+                if not len(self.current_batch) and num_new_batched_tokens > self.max_num_batched_tokens:
+                    state.token_ids = state.token_ids[:self.max_num_batched_tokens]
+                    state.next_start_position = num_new_batched_tokens = num_tokens = self.max_num_batched_tokens
                 if num_new_batched_tokens > self.max_num_batched_tokens > 0:
                     LOG.debug(
                         "Stop growing the batch due to max_num_batched_tokens. Batched tokens: %s",
                         num_new_batched_tokens,
                     )
                     break
                 if (
-                    self.cache_manager.get_free_space()
-                    <= self.prompt_allocate_ratio * num_tokens
+                    (self.cache_manager.get_free_space() - num_tokens) < 0 or \
+                         (self.cache_manager.get_free_space() - num_tokens) / (len(self.current_batch) + 1) < self.max_decode_steps
```

Comment:
Seems `(self.cache_manager.get_free_space() - num_tokens) < 0` is redundant? It is subsumed by the second condition.

Reply:
fixed